File: src/25_Reverse_Nodes_in_k-Group.c
// https://leetcode.cn/problems/reverse-nodes-in-k-group/

// 8 ms 69.75%
// 6.8 ms 77.02%

#include<stdio.h>

// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode *pre = NULL;
    struct ListNode *cur = head;
    struct ListNode *tmp = NULL;

    while (cur) {
        tmp = cur->next;
        cur->next = pre;
        pre = cur;
        cur = tmp;
    }

    return pre;
}


struct ListNode* reverseKGroup(struct ListNode* head, int k){
    struct ListNode dummy;
    dummy.next = head;
    int len = get_length(dummy.next);
    int times = len / k;

    struct ListNode *before = &dummy;
    for (int i=0; i<times; i++) {
        struct ListNode *left_node = before->next;
        struct ListNode *right_node = before;
        for (int j=0; j<k; j++) {
            right_node = right_node->next;
        }
        struct ListNode *succ = right_node->next;

        before->next = NULL;
        right_node->next = NULL;

        reverseList(left_node);

        before->next = right_node;
        left_node->next = succ;

        for (int m=0; m<k; m++) {
            before = before->next;
        }
    }

    return dummy.next;
}


int get_length(struct ListNode* head) {
    int len = 0;
    while (head) {
        len++ ;
        head = head->next;
    }

    return len;
}

File: src/25_Reverse_Nodes_in_k-Group.rs
// https://leetcode.cn/problems/reverse-nodes-in-k-group/

// 0 ms 100%
// 2,2 MB 82.93%

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode {
      next: None,
      val
    }
  }
}


// 本质就是某个节点最多只有一个可变引用问题，不然代码里面就没有必要迭代每段翻转后的一组链表找到它的尾结点当下一组的前驱
impl Solution {
    pub fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut remain = head;
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;
        while remain.is_some() {
            let (new_head, new_remain) = Solution::reverse_one(remain, k);
            remain = new_remain;
            tail.next = new_head;
            while tail.next.as_ref().is_some() {
                tail = tail.next.as_mut().unwrap();
            }
        }

        dummy.next
    }

    // 反转一次，返回反转后的head和remain
    // 如果为最后一次不足以反转，remain为None
    fn reverse_one(head: Option<Box<ListNode>>, k: i32) -> (Option<Box<ListNode>>, Option<Box<ListNode>>) {
        let mut pre = head.as_ref();
        for _ in 0..k {
            if pre.is_none() {
                return (head, None);
            }
            pre = pre.unwrap().next.as_ref();
        }

        let mut remain = head;
        let mut dummy = ListNode::new(0);
        for _ in 0..k {
            if let Some(mut n) = remain {
                remain = n.next.take();
                n.next = dummy.next.take();
                dummy.next = Some(n);
            }
        }

        (dummy.next, remain)
    }
}
// 作者：psy-core
// 链接：https://leetcode.cn/problems/reverse-nodes-in-k-group/solution/rustlian-biao-cao-zuo-by-psy-core-o6j3/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。